pd/src/components/staking.rs

```diff
@@ -354,16 +369,59 @@ pub trait View: WriteOverlayExt + Send + Sync {
         self.validator(&identity_key).await
     }
 
-    async fn slash_validator(&mut self, evidence: &Evidence, slashing_penalty: u64) -> Result<()> {
+    async fn slash_validator(&mut self, evidence: &Evidence) -> Result<()> {
         let ck = tendermint::PublicKey::from_raw_ed25519(&evidence.validator.address)
             .ok_or_else(|| anyhow::anyhow!("invalid ed25519 consensus pubkey from tendermint"))
             .unwrap();
 
-        let validator = self.validator_by_consensus_key(&ck).await?;
+        let validator = self
+            .validator_by_consensus_key(&ck)
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("attempted to slash validator not found in JMT"))?;
+
+        let slashing_penalty = self.get_chain_params().await?.slashing_penalty;
 
         tracing::info!(?validator, ?slashing_penalty, "slashing validator");
 
-        // TOD: implement slashing
+        let cur_state = self
+            .validator_state(&validator.identity_key)
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("validator to be slashed did not have state in JMT"))?;
+
+        // Ensure that the state transitions are valid.
+        match cur_state {
+            ValidatorState::Active => {}
+            ValidatorState::Unbonding { unbonding_epoch: _ } => {}
+            _ => {
+                return Err(anyhow::anyhow!(
+                    "only validators in the active or unbonding state may be slashed"
+                ))
+            }
+        };
+
+        // Mark the state as "slashed" in the JMT, and apply the slashing penalty.
+        self.set_validator_state(&validator.identity_key, ValidatorState::Slashed);
+
+        let mut cur_rate = self
+            .current_validator_rate(&validator.identity_key)
+            .await?
+            .ok_or_else(|| {
+                anyhow::anyhow!("validator to be slashed did not have current rate in JMT")
+            })?;
+
+        cur_rate = cur_rate.slash(slashing_penalty);
+
+        // TODO: would it be better to call `current_base_rate.next`? the same logic exists
+        // within there, but it requires passing in the current base rates & funding streams,
+        // which aren't actually used because the rate is held constant. So, doing it this way
+        // avoids a couple unnecessary JMT reads that the `current_base_rate.next` API would require.
+        //
+        // At any rate, the next rate is held constant for slashed validators.
```

Comment:
I think it's better not to call `next` here, since I think that will evolve the rates through time, but we know that the next rate should be constant.

Reply:
The implementation of `next` is state-aware and will hold a constant rate for a slashed validator, since `next` is what's called for all validators during `end_epoch`.

Reply:
Ah, great, never mind then!  I think just bumping up the epoch here is simpler (from the reader's point of view) than having to fetch a bunch of extra data.